src/main/scala/io/iohk/ethereum/blockchain/sync/PeersClient.scala

```diff
@@ -129,6 +136,25 @@ class PeersClient(
 
     log.debug(s"Handshaked peers status (number of peers: ${handshakedPeersStatus.size}): $handshakedPeersStatus")
   }
+
+  //returns the next best peer after the one already returned previously
+  //TODO: make sure the next best peer has a different best block, so we don't fetch in parallel identical branches
+  private def nextBestPeer(peersToDownloadFrom: Map[PeerId, PeerWithInfo]): Option[Peer] = {
+    val peersToUse = peersToDownloadFrom.values
+      .collect { case PeerWithInfo(peer, PeerInfo(_, chainWeight, true, _, _)) =>
+        (peer, chainWeight)
+      }
+
+    val peer =
+      peersToUse.filter(peerAndWeight => !activeFetchingNodes.contains(peerAndWeight._1)).maxByOption(_._2).map(_._1)
+    peer.foreach(activeFetchingNodes.add)
```

Comment:
I think that hiding side effects makes the code harder to read and reason about.